repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/RequestHandler.scala

```diff
@@ -0,0 +1,524 @@
+ *
+ * Scoped to a single request/thread.
+ */
+class RequestHandler(config: KeystoneV2Config, akkaServiceClient: AkkaServiceClient, datastore: Datastore, traceId: Option[String])
+  extends LazyLogging {
+
+  import RequestHandler._
+
+  final def validateToken(authenticatingToken: String, token: String): Try[AuthResult] = {
+    /**
+     * Extract the user's information from the validate token response
+     * @param inputStream the validate token response!
+     * @return a success or failure of ValidToken information
+     */
+    def extractUserInformation(inputStream: InputStream): Try[ValidToken] = {
+      import play.api.libs.json.Reads._
+      import play.api.libs.json._
+
+      val input: String = Source.fromInputStream(inputStream).getLines mkString ""
+      try {
+        val json = Json.parse(input)
+        //Have to convert it to a vector, because List isn't serializeable in 2.10
+        val roleNames = (json \ "access" \ "user" \ "roles" \\ "name").map(_.as[String]).toVector
+        val defaultTenantId = (json \ "access" \ "token" \ "tenant" \ "id").as[String]
+        val tenantIds = (json \ "access" \ "user" \ "roles" \\ "tenantId").map(_.as[String]).toVector
+        val userId = (json \ "access" \ "user" \ "id").as[String]
+        // note: this may be optional? if so, asOpt can be used.
+        val username = (json \ "access" \ "user" \ "name").as[String]
+        val tenantName = (json \ "access" \ "token" \ "tenant" \ "name").as[String]
+        val defaultRegion = (json \ "access" \ "user" \ "RAX-AUTH:defaultRegion").asOpt[String]
+        val contactId = (json \ "access" \ "user" \ "RAX-AUTH:contactId").asOpt[String]
+        val expirationDate = iso8601ToRfc1123((json \ "access" \ "token" \ "expires").as[String])
+        val impersonatorId = (json \ "access" \ "RAX-AUTH:impersonator" \ "id").asOpt[String]
+        val impersonatorName = (json \ "access" \ "RAX-AUTH:impersonator" \ "name").asOpt[String]
+        val validToken = ValidToken(expirationDate, userId, username, tenantName, defaultTenantId, tenantIds, roleNames, impersonatorId, impersonatorName, defaultRegion, contactId)
+
+        Option(config.getCache) foreach { cacheSettings =>
+          // never null because configurationUpdated in KeystoneV2Filter
+          val timeout = offsetTtl(cacheSettings.getTimeouts.getToken, cacheSettings.getTimeouts.getVariability)
+          datastore.put(s"$TOKEN_KEY_PREFIX$token", validToken, timeout, TimeUnit.SECONDS)
+        }
+        Success(validToken)
+      } catch {
+        case oops@(_: JsResultException | _: JsonProcessingException) =>
+          Failure(IdentityCommuncationException("Unable to parse JSON from identity validate token response", oops))
+      }
+    }
+
+    val identityEndpoint = config.getIdentityService.getUri
+
+    import scala.collection.JavaConverters._
+    Try(akkaServiceClient.get(s"$TOKEN_KEY_PREFIX$token",
+      s"$identityEndpoint$TOKEN_ENDPOINT/$token",
```

Comment:
I'm going to skip this one for now because I can't find a URI Builder that does what I want. It looks like the javax URIBuilder will, but trying to test the usage of that class leads to a ClassNotFoundException.